apps/prairielearn/assets/scripts/instructorAssessmentAccessClient.ts

```diff
@@ -0,0 +1,225 @@
+    const endDate = adjustedDate(updatedAccessRules.end_date.toString());
+    updatedAccessRules.start_date =
+      formatDate(startDate, 'UTC', { includeTz: false }) + ` (${timeZoneName})`;
+    updatedAccessRules.end_date =
+      formatDate(endDate, 'UTC', { includeTz: false }) + ` (${timeZoneName})`;
+    accessRulesData[row] = updatedAccessRules;
+    refreshTable();
+  }
+
+  on('click', '.editButton', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    const accessRule = accessRulesData[row];
+    accessRule.formatted_start_date = adjustedDate(accessRule.start_date)
+      .toISOString()
+      .slice(0, 19);
+    accessRule.formatted_end_date = adjustedDate(accessRule.end_date).toISOString().slice(0, 19);
+    if (!timeZoneName) {
+      throw new Error('Course instance time zone is not available.');
+    }
+    morphdom(
+      editAccessRuleModal as Node,
+      EditAccessRuleModal({ accessRule, i: row, timeZoneName }).toString(),
+    );
+    $('#editAccessRuleModal').modal('show');
+  });
+
+  on('click', '#updateAccessRuleButton', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    const form = (e.target as HTMLElement).closest('form');
+    if (!form) return;
+    handleUpdateAccessRule({ form, row });
+  });
+
+  function swapRows(row: any, targetRow: any) {
+    const accessRule = accessRulesData[row];
+    accessRulesData[row] = accessRulesData[targetRow];
+    accessRulesData[targetRow] = accessRule;
+    refreshTable();
+  }
+
+  on('click', '.up-arrow-button', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    if (!row) return;
+    swapRows(row, row - 1);
+  });
+
+  on('click', '.down-arrow-button', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    if (row === accessRulesData.length - 1) return;
+    swapRows(row, row + 1);
+  });
+
+  on('click', '#addRuleButton', () => {
+    const formatted_start_date = adjustedDate(new Date()).toISOString().slice(0, 19);
+    const formatted_end_date = adjustedDate(new Date()).toISOString().slice(0, 19);
+    if (!timeZoneName) {
+      throw new Error('Course instance time zone is not available.');
+    }
+    morphdom(
+      editAccessRuleModal as Node,
+      EditAccessRuleModal({
+        accessRule: {
+          mode: '',
+          uids: '',
+          start_date: new Date().toISOString().slice(0, 19),
+          end_date: new Date().toISOString().slice(0, 19),
+          active: 'True',
+          credit: '100',
+          time_limit: '50',
+          password: '',
+          formatted_start_date,
+          formatted_end_date,
+          exam_uuid: null,
+          pt_course_id: null,
+          pt_course_name: null,
+          pt_exam_id: null,
+          pt_exam_name: null,
+        },
```

Comment:
I have removed defaults for most of these with the exception of `start_date`, `end_date` and `active`. `start_date` and `end_date` will throw errors if they are blank so I have them default to the current time. `active` is a boolean and needs to be either `true` or `false`.  

Also, because of the changes in [1dfef07](https://github.com/PrairieLearn/PrairieLearn/pull/9934/commits/1dfef07527ed0296ff3cdfa7c58d85fce2304b97), the default values are now set from the `addRuleButton` on the `accessRulesTable`.

Reply:
> `start_date` and `end_date` will throw errors if they are blank

What exactly will throw errors? There are certainly valid cases to have blank start/end dates, namely PrairieTest access rules.

Reply:
The errors were coming from an empty object when passing it back to the client. I have now added conditionals to handle blank start/end dates and updated the defaults for these in a new rule to be blank.